### ISSUE TITLE
Error: Unable to load main class org.getmonero.i2p.zero.Main in module org.getmonero.i2p.zero

### ISSUE THREAD

**Issue by MoneroArbo** [knaccc/i2p-zero#31]
I get an error when starting the Linux v1.18 release:

`Error: Unable to load main class org.getmonero.i2p.zero.Main in module org.getmonero.i2p.zero
        java.lang.UnsupportedClassVersionError: Preview features are not enabled for org/getmonero/i2p/zero/Main (class file version 58.65535). Try running with '--enable-preview'`

Running with "--enable-preview" gives the same error.

I was previously running v1.17 with no issues. After reverting back to v1.17 everything works again.

OS is Ubuntu 20.04

**Comment by knaccc** [comment]
Thanks for the bug report. Please try `--enable-preview -source 14`. Of course, you shouldn't have to do that - I'll look into fixing this and releasing a new version.

**Comment by MoneroArbo** [comment]
Hey thanks for the response! Unfortunately it gives the same error. I tried with `--enable-preview -source 14` as well as `--enable-preview --source 14`.

Edit just to add, I'm fine with running the old version until a fix can be released. Still wanted to report back with my results though, in case it's helpful.

**Comment by knaccc** [comment]
Btw it needs to be added at the beginning and not the end. I.e. your launch.sh file should contain:

`"$basedir"/bin/java --enable-preview -source 14 -cp "$basedir/i2p.base/jbigi.jar" -m org.getmonero.i2p.zero`

**Comment by MoneroArbo** [comment]
Sorry, I'm a little confused by these instructions since I don't have Java installed. I've never needed it for previous versions and my understanding of i2p-zero is it doesn't require any dependencies, so... I could of course install Java, but I think maybe there's a misunderstanding.

Would it help if I specified that I am NOT running the Docker build? I'm running the Linux v1.18 release with sha256sum `7901bde9dfc6e75adccfe304a1abb057a5e3c795cec74fdce61d52251aa151ac`

**Comment by MoneroArbo** [comment]
Okay sorry it took me awhile to figure out that basedir=router/lib/runtime -- I was initially looking in router/bin

But after sorting that out I got: `Error: Option -m is not allowed with --source`

I also tried `-source` but it was not recognized as a valid option.

However, trying the command with no `--source` option:

` "$basedir"/bin/java --enable-preview -cp "$basedir/i2p.base/jbigi.jar" -m org.getmonero.i2p.zero`

caused it to launch and it now appears to be running with no issues.

I have modified my start script appropriately.

Thanks for the help!

**Comment by knaccc** [comment]
@MoneroArbo Thanks for letting me know

**Comment by re42** [comment]
Hi, having the same problem on Xubuntu 18.04.

So i tried to run this command from home:
`java --enable-preview -cp "/home/user/i2p-zero-linux-gui.v1.18/router/lib/runtime/i2p.base/jbigi.jar" -m org.getmonero.i2p.zero`

But i get this error:
```
Error occurred during initialization of boot layer
java.lang.module.FindException: Module org.getmonero.i2p.zero not found
```

Unless im missing a step in the process of running the i2p Linux GUI, i dont seem to understand where org.getmonero.i2p.zero is situated and why this wasnt a problem in 1.17?
Please help

**Comment by MoneroArbo** [comment]
Maybe you need to use the java binary that ships with i2p-zero? It's located at `router/lib/runtime/bin/java`

**Comment by re42** [comment]
> Maybe you need to use the java binary that ships with i2p-zero? It's located at `router/lib/runtime/bin/java`

Thank you @MoneroArbo, that worked, but im using the i2p-zero Linux **GUI**, and when i ran that command, it only executed correctly in terminal, but im not sure how to get the GUI running?  i replaced `"/home/user/i2p-zero-linux-gui.v1.18/router/lib/runtime/i2p.base/jbigi.jar"` with `"/home/user/i2p-zero-linux-gui.v1.18/router/bin/i2p-zero"`  but still no GUI.
any ideas?

**Comment by re42** [comment]
Ok so i figured out i had to use the module for the GUI, which is `org.getmonero.i2p.zero.gui`  but i am having another problem now.
I ran:
`"/home/user/i2p-zero-linux-gui.v1.18/router/lib/runtime/bin/java" --enable-preview -cp "/home/user/i2p-zero-linux-gui.v1.18/router/lib/runtime/i2p.base/jbigi.jar" -m org.getmonero.i2p.zero.gui`

and i get the error:
```
ES2 Prism: Error - reported GLX version = 1.2
    GLX version 1.3 or higher is required
Exception in Application start method
Exception in Application stop method
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplicationWithArgs(LauncherImpl.java:464)
	at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication(LauncherImpl.java:363)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at java.base/sun.launcher.LauncherHelper$FXHelper.main(LauncherHelper.java:1051)
Caused by: java.lang.RuntimeException: Exception in Application start method
	at javafx.graphics/com.sun.javafx.application.LauncherImpl.launchApplication1(LauncherImpl.java:900)
	at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication$2(LauncherImpl.java:195)
	at java.base/java.lang.Thread.run(Thread.java:832)
Caused by: javafx.fxml.LoadException: 
/org.getmonero.i2p.zero.gui/org/getmonero/i2p/zero/gui/gui.fxml

	at javafx.fxml/javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2625)
	at javafx.fxml/javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2595)
	at javafx.fxml/javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2466)
	at javafx.fxml/javafx.fxml.FXMLLoader.load(FXMLLoader.java:2435)
	at org.getmonero.i2p.zero.gui/org.getmonero.i2p.zero.gui.Gui.start(Gui.java:46)
	at javafx.graphics/com.sun.javafx.application.LauncherImpl.lambda$launchApplication1$9(LauncherImpl.java:846)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runAndWait$12(PlatformImpl.java:455)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$11(GtkApplication.java:277)
	... 1 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:76)
	at jdk.internal.reflect.GeneratedMethodAccessor2.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at javafx.base/com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:273)
	at javafx.fxml/com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:83)
	at javafx.fxml/javafx.fxml.FXMLLoader.loadImpl(FXMLLoader.java:2591)
	... 12 more
Caused by: java.lang.RuntimeException: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
	at org.getmonero.i2p.zero/org.getmonero.i2p.zero.RouterWrapper.loadBandwidthLimitKBps(RouterWrapper.java:186)
	at org.getmonero.i2p.zero.gui/org.getmonero.i2p.zero.gui.Controller.initialize(Controller.java:309)
	... 23 more
Caused by: java.lang.ClassCastException: class java.lang.Integer cannot be cast to class java.lang.Long (java.lang.Integer and java.lang.Long are in module java.base of loader 'bootstrap')
	at org.getmonero.i2p.zero/org.getmonero.i2p.zero.RouterWrapper.loadBandwidthLimitKBps(RouterWrapper.java:182)
	... 24 more
Exception running application org.getmonero.i2p.zero.gui.Gui
```
This error is specific to the i2p-zero Linux GUI.  Any idea why its throwing this error @knaccc ?

(Bear in mind, that i used to get this specific error message `ES2 Prism: Error - reported GLX version = 1.2    GLX version 1.3 or higher is required` in 1.17 as well, but the GUI still ran perfectly fine)  So i take it this has nothing to do with GLX version?

**Comment by knaccc** [comment]
@re42 It looks like I messed up the v1.18 release because my test environment was set up with certain defaults that the release did not set. I'll see if I can get it right for the upcoming v1.19 release.

Are you saying that v1.17 worked perfectly for you (i.e. errors may have displayed, but nothing was preventing the software from working properly?)

**Comment by re42** [comment]
@knaccc yes im using v1.17 right now, and it works perfectly fine, as it always did (Even though i see the GLX error when executing it)

**Comment by knaccc** [comment]
FYI I'm running into two problems here. Firstly, there seem to be problems getting --enable-preview to work properly when things are packaged for cross-platform release. I think this is a bug in the JDK jpackage system.

As well as that, I was going to try and fix this along with the latest JDK15 release, but that seems to have totally broken jlink/jpackage (upon which the cross-platform stuff depends). I'm hoping there will be a bugfix for JDK15. If not, the solution will probably be to revert to JDK14 and remove the code that relies on the new Java 14 "Records" preview feature.

Therefore please use v1.17 while this is being resolved.

**Comment by Engelberg** [comment]
I'm having the same problem. Following so I see updates to this issue.

**Comment by knaccc** [comment]
Fixed in v1.19 just released
